### ISSUE TITLE
False Negative: The rule focuses too narrowly on string literals and misses `==` comparisons involving variables and `new String(...)`.

### ISSUE THREAD

**Issue by Carlson-JLQ** [checkstyle/checkstyle#19390]
# False Negative: The rule focuses too narrowly on string literals and misses `==` comparisons involving variables and `new String(...)`.

Version
unknown in repository metadata

## Checker
- Checker id: `StringLiteralEqualityCheck`
- Checker description: Detects equality operators (== or !=) that have a string literal operand in their expression subtree.

## Description of the false negative
`diff/output` marks 2 false negatives for `StringLiteralEqualityCheck`. Across these samples, the benchmark is still exercising the same underlying bug pattern, but the implementation appears to lose track of it once the source deviates from one canonical shape. The recurring themes are string reference equality.

### `PosCase1.java`
Source: `output/codeql/StringComparison/src/main/java/scensct/core/pos/PosCase1.java`

The core operation is a reference-equality check involving strings. The misuse is still a plain reference comparison on string values, which should not depend on whether one operand is a literal or a variable. That is also how the benchmark classifies the sample: Comparing a new String object with a string literal using == should be flagged as a reference equality comparison with a non-interned operand.

```java
// Comparing a new String object with a string literal using == should be flagged as a reference equality comparison with a non-interned operand.
package scensct.core.pos;

public class PosCase1 {
    public static void main(String[] args) {
        String literal = "hello";
        String newString = new String("hello");
        // Scenario: new String("literal") is not guaranteed to be interned, compared with ==
        boolean result = (newString == literal); // Should trigger report // [REPORTED LINE]
        System.out.println(result);
    }
}
```

### `PosCase1.java`
Source: `output/pmd/UseEqualsToCompareStrings/src/main/java/scensct/core/pos/PosCase1.java`

The core operation is a reference-equality check involving strings. The misuse is still a plain reference comparison on string values, which should not depend on whether one operand is a literal or a variable. That is also how the benchmark classifies the sample: Compare two String variables using the '==' operator should be flagged as string comparison with '=='.

```java
// Compare two String variables using the '==' operator should be flagged as string comparison with '=='.
package scensct.core.pos;

public class PosCase1 {
    public static void main(String[] args) {
        String str1 = "hello";
        String str2 = "world";
        // Direct comparison of two String variables using '=='
        boolean result = (str1 == str2); // [REPORTED LINE]
        System.out.println(result);
    }
}
```

## Cause analysis
The rule is documented as follows: Detects equality operators (== or !=) that have a string literal operand in their expression subtree.

In this set the problem shows up around string reference equality. The missed cases suggest that the implementation is narrower than the rule description and too dependent on one preferred AST shape.

Once the same behavior is expressed through a helper, an overload, a modifier such as `abstract` or `native`, a different comment position, or a slightly rearranged control-flow structure, the report disappears even though the benchmark still treats the case as in-scope.

For `StringLiteralEqualityCheck`, the practical improvement would be to generalize the match so it follows the underlying behavior rather than one exact spelling of it.

## References
None known.

**Comment by romani** [comment]
please follow issue template
